packages/SwingSet/src/kernel/kernel.js

```diff
@@ -503,23 +510,93 @@ export default function buildKernel(kernelEndowments) {
-      vatID,
-      harden({
-        manager,
-        enablePipelining: Boolean(options.enablePipelining),
-      }),
+      kernelKeeper.provideVatKeeper(vatID),
     );
   }
 
+  // This kernel (and the genesis vats) will have been evaluated in a context in
+  // which these canonical symbols were available (which is why the require
+  // statements here work). That was achieved by running makeRequire() in
+  // controller.js. We want to provide the same context when we evaluate dynamic
+  // vats.
+  function kernelRequire(nameArg) {
+    const name = `${nameArg}`;
+    const HARDEN = '@agoric/harden';
+    const NAT = '@agoric/nat';
+    const EVALUATE = '@agoric/evaluate';
+    switch (name) {
+      case HARDEN:
+        // eslint-disable-next-line global-require,import/no-dynamic-require
+        return require(HARDEN);
+      case NAT:
+        // eslint-disable-next-line global-require,import/no-dynamic-require
+        return require(NAT);
+      case EVALUATE:
+        // eslint-disable-next-line global-require,import/no-dynamic-require
+        return require(EVALUATE);
+      default:
+        throw Error(`require ${name} is not supported`);
+    }
+  }
+
+  // Enqueue a message to the adminVat giving it the new vat's root object
+  function notifyAdminVatOfNewVat(vatID) {
+    function serializeSlot(slot) {
+      return {
+        body: `["${vatID}",{"@qclass":"slot","index":0}]`,
+        slots: [slot],
+      };
+    }
+    const vatSlot = makeVatRootObjectSlot();
+    const kernelRootObjSlot = addExport(vatID, vatSlot);
+    const serializedArgs = serializeSlot(kernelRootObjSlot);
+    const vatAdminVatId = vatNameToID('vatAdmin');
+    queueToExport(vatAdminVatId, vatSlot, 'newVatCallback', serializedArgs);
+  }
+
+  // Create a new vat and return the vatID without waiting for the root object.
+  function createVat(buildFn) {
+    const vatID = kernelKeeper.provideUnusedVatID();
+
+    const setup = (syscall, state, helpers) => {
+      return helpers.makeLiveSlots(syscall, state, buildFn, helpers.vatID);
+    };
+
+    let manager;
+    try {
+      manager = buildVatManager(vatID, `dynamicVat${vatID}`, setup);
+    } catch (e) {
+      return `${e}`;
+    }
+    ephemeral.vats.set(vatID, harden({ manager }));
+    return vatID;
+  }
+
+  // A function to be called from the vatAdmin device to create a new vat. It
+  // creates the vat and sends a notification to the device. The root object
+  // will be available soon, but we immediately return the vatID so the ultimate
+  // requestor doesn't have to wait.
+  function createVatDynamically(buildFnSrc) {
+    // use kernelRequire to get the evaluate we want.
+    const kernelEvaluate = kernelRequire('@agoric/evaluate');
+    // pass kernelRequire to evaluate for buildFn's use
+    const req = { require: kernelRequire };
+    const buildFn = kernelEvaluate.evaluateProgram(buildFnSrc, req);
+    const vatID = createVat(buildFn);
+
+    if (`${vatID}`.startsWith('v')) {
```

Comment:
when is this not true?

Reply:
We discussed this in person.. turns out `createVat` has an obvious error case (when the user-supplied function string fails to parse or throws an exception as it creates the root object). The plan is to let `createVat` raise the exception, but catch it here and notify the manager vat via two different messages: one for success (with the new vatID) and the other for failure (with an error string), both of which are used to resolve/reject the originating vat's request promise.

Reply:
ok I now think I understand how this works. But I'm not excited about the API: the caller who invokes `createVat()` should get back a promise that either resolves to a `{ adminNode, root }` or rejects to an error. Having it always resolve (sometimes to `{ adminNode, root }` and sometimes to `"Error:..."`) feels fragile. Let's land this as-is, and clean it up in a second pass. I've added #445 to remind us.